core/solver/gmres.cpp

```diff
@@ -312,36 +504,39 @@ void Gmres<ValueType>::apply_dense_impl(const VectorType* dense_b,
         this->get_preconditioner()->apply(this_krylov,
                                           preconditioned_krylov_vector);
 
-        // Create view of current column in the hessenberg matrix:
-        // hessenberg_iter = hessenberg(:, restart_iter);
-        auto hessenberg_iter = hessenberg->create_submatrix(
-            span{0, restart_iter + 2},
-            span{num_rhs * restart_iter, num_rhs * (restart_iter + 1)});
+        // Create view of current "column" in the hessenberg matrix:
+        // hessenberg_iter = hessenberg(:, restart_iter), which
+        // is actually stored as a row, hessenberg(restart_iter, :)
+        auto hessenberg_iter =
+            hessenberg->create_submatrix(span{restart_iter, restart_iter + 1},
+                                         span{0, num_rhs * (restart_iter + 2)});
```

Comment:
If I understand the storage layout correctly, you could 'reshape' this row (aka column) into a `N_basis x N_rhs` matrix, right? Then some of the changes related to the new indexing might not be necessary. But I've not checked it fully to see how much it would actually help-

Reply:
Do you mean by directly creating a `Dense` of shape `[restart_iter + 2, num_rhs]` with a view to the correctly offset starting point in `hessenberg`'s values for the data? That would mean the `hessenberg_qr` kernel could revert back to its original version, but `solve_krylov` would still need to be altered since it considers the whole Hessenberg matrix. 

I think it depends on whether we think it's more or less clear to "mix" the two in this way -- meaning, if the full Hessenberg matrix is viewed in its new layout, while the "for-this-iteration" slice of it is viewed in the "logical" way, with rows corresponding to iterations and columns to the rhs vector. Viewing `hessenberg_iter` in the "logical" way would probably make routines only involving `hessenberg_iter` more clear, but then is it more confusing overall? I could see the argument either way.

Reply:
I think it would be best to have both views available. They are free anyway, and I think if we name them properly they should be a net benefit.
The most natural way would probably be to just create the view on-the-fly, but this doesn't work with our separation of our core and device libraries. So we need to create the views beforehand.

Reply:
Why would we need both views? What would we do with each?

Reply:
Oh, I guess I misunderstood your comment. So again, the question is, which view is more useful for kernels, but I think you can judge that better. IMO, using indexing like `hess(i, j)` seems easier to me than `hess(0, i * N + j)` (or similar).

Reply:
In 35e965c065fc190dcef09278a62741aca8afad4e I changed the `hessenberg_iter` layout.